python_modules/dagster/dagster/_core/definitions/observable_asset.py

```diff
@@ -79,8 +86,17 @@ def create_unexecutable_observable_assets_def_from_source_asset(source_asset: So
     elif source_asset.io_manager_key:
         kwargs["io_manager_key"] = source_asset.io_manager_key
 
+    kwargs["partitions_def"] = source_asset.partitions_def
+
+    if source_asset.observe_fn:
+        kwargs["resource_defs"] = source_asset.resource_defs
+
     @asset(**kwargs)
-    def shim_asset() -> None:
-        raise NotImplementedError(f"Asset {source_asset.key} is not executable")
+    def shim_asset(context: OpExecutionContext) -> None:
+        if not source_asset.observe_fn:
+            raise NotImplementedError(f"Asset {source_asset.key} is not executable")
+
+        op_function = wrap_source_asset_observe_fn_in_op_compute_fn(source_asset)
+        return op_function.decorated_fn(context)
```

Comment:
what happens output wise for observable source assets? Does the default io manager store whatever dataversion object gets returned or do we have special casing somewhere?

Reply:
Confirmed that this does not result in the emission of an Output.

1) `DataVersion` gets translated to an `AssetObservation` in `wrap_source_asset_observe_fn_in_op_compute_fn`
2) `wrap_source_asset_observe_fn_in_op_compute_fn` also creates a compute fn guaranteed to return None.

2 is not enforced by the type system, but I have added an invariant with a helpful message to describe the state of things

Also added a test to confirm this behavior.